sources/web/datalab/polymer/components/datalab-editor/datalab-editor.ts

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright 2017 Google Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+/**
+ * Type declarations for CodeMirror
+ */
+
+interface CodeMirrorOptions {
+  value: string,
+  mode: string,
+  lineNumbers: boolean,
+  lineWrapping: boolean,
+  theme: string,
+}
+
+interface CodeMirrorEditor {
+  setOption(name: string, value: string): void;
+  getOption(name: string): void;
+}
+
+declare function CodeMirror (element: HTMLElement, options: CodeMirrorOptions): CodeMirrorEditor;
+
+/**
+ * Editor element for Datalab.
+ * Contains a <datalab-toolbar> element at the top, and a full screen editor
+ * that uses CodeMirror.
+ */
+class DatalabEditorElement extends Polymer.Element {
+
+  /**
+   * Path of the file to load in the editor.
+   */
+  public filePath: string;
+
+  private _editor: CodeMirrorEditor;
+  private _theme: string;
+  private _busy: boolean;
+
+  static get is() { return 'datalab-editor'; }
+
+  static get properties() {
+    return {
+      filePath: {
+        type: String,
+        value: '',
+      },
+      _busy: {
+        type: Boolean,
+        value: false
+      },
+    }
+  }
+
+  ready() {
+    super.ready();
+
+    // Get the theme.
+    SettingsManager.getUserSettingsAsync()
+      .then((settings: common.UserSettings) => {
+        if (settings.theme) {
+          this._theme = settings.theme;
+        }
+      })
+      // Get the file contents, or empty string if no path is specified.
+      .then(() => {
+        if (this.filePath !== '') {
+          this._busy = true;
+          return ApiManager.getJupyterFile(this.filePath, true /*asText*/)
+            .then((file: JupyterFile) => file.content)
+            .catch((e: Error) => {
+              // TODO: Handle error visibly to the user.
+              console.log('Could not load specified file: ', e);
+              return '';
+            })
+        } else {
+          return '';
+        }
+      })
+      // Create the codemirror element and load the contents in it.
+      .then((content: string) => {
+        // TODO: try to detect the language of the file before creating
```

Comment:
Can you pass the filename in here as well and use that to help determine the file type?